c/tskit/trees.c

```diff
@@ -5485,164 +5635,73 @@ tsk_tree_insert_edge(tsk_tree_t *self, tsk_id_t p, tsk_id_t c, tsk_id_t edge_id)
+    ret = tsk_tree_clear(self);
+    if (ret != 0) {
+        goto out;
     }
+    ret = tsk_tree_prev(self);
 out:
     return ret;
 }
 
-int TSK_WARN_UNUSED
-tsk_tree_last(tsk_tree_t *self)
+static void
+tsk_tree_update_index_and_interval(tsk_tree_t *self)
 {
-    int ret = TSK_TREE_OK;
-    const tsk_treeseq_t *ts = self->tree_sequence;
-    const tsk_table_collection_t *tables = ts->tables;
+    tsk_table_collection_t *tables = self->tree_sequence->tables;
 
-    self->interval.left = 0;
-    self->interval.right = tables->sequence_length;
-    self->index = 0;
-    self->sites = ts->tree_sites[0];
-    self->sites_length = ts->tree_sites_length[0];
-
-    if (tables->edges.num_rows > 0) {
-        /* TODO this is redundant if this is the first usage of the tree. We
-         * should add a state machine here so we know what state the tree is
-         * in and can take the appropriate actions.
-         */
-        ret = tsk_tree_clear(self);
-        if (ret != 0) {
-            goto out;
-        }
-        self->index = (tsk_id_t) tsk_treeseq_get_num_trees(ts);
-        self->left_index = (tsk_id_t) tables->edges.num_rows - 1;
-        self->right_index = (tsk_id_t) tables->edges.num_rows - 1;
-        self->direction = TSK_DIR_REVERSE;
-        self->interval.left = tables->sequence_length;
-        self->interval.right = 0;
+    self->index = self->tree_pos.index;
+    self->interval.left = self->tree_pos.interval.left;
+    self->interval.right = self->tree_pos.interval.right;
 
-        ret = tsk_tree_advance(self, TSK_DIR_REVERSE, tables->edges.left,
-            tables->indexes.edge_insertion_order, &self->left_index, tables->edges.right,
-            tables->indexes.edge_removal_order, &self->right_index);
+    if (tables->sites.num_rows > 0) {
+        self->sites = self->tree_sequence->tree_sites[self->index];
+        self->sites_length = self->tree_sequence->tree_sites_length[self->index];
     }
-out:
-    return ret;
 }
 
 int TSK_WARN_UNUSED
 tsk_tree_next(tsk_tree_t *self)
 {
     int ret = 0;
-    const tsk_treeseq_t *ts = self->tree_sequence;
-    const tsk_table_collection_t *tables = ts->tables;
-    tsk_id_t num_trees = (tsk_id_t) tsk_treeseq_get_num_trees(ts);
+    tsk_table_collection_t *tables = self->tree_sequence->tables;
+    const tsk_id_t *restrict edge_parent = tables->edges.parent;
+    const tsk_id_t *restrict edge_child = tables->edges.child;
+    tsk_id_t j, e;
+    bool valid;
 
-    if (self->index == -1) {
-        ret = tsk_tree_first(self);
-    } else if (self->index < num_trees - 1) {
-        ret = tsk_tree_advance(self, TSK_DIR_FORWARD, tables->edges.right,
-            tables->indexes.edge_removal_order, &self->right_index, tables->edges.left,
-            tables->indexes.edge_insertion_order, &self->left_index);
+    valid = tsk_tree_position_next(&self->tree_pos);
+
+    if (valid) {
+        for (j = self->tree_pos.out.start; j != self->tree_pos.out.stop; j++) {
```

Comment:
I think these pointer references might be adding up. Try something like
```
tree_pos = self->tree_pos;
for (j = tree_pos.out.start; j != tree_pos.out.stop; j++) {
...
}```

Reply:
Looks like this fixes the problem. I'm just running a better perftest script now to make sure